core/src/main/scala/com/pingcap/tispark/telemetry/TiSparkTeleInfo.scala

```diff
@@ -0,0 +1,157 @@
+
+import com.pingcap.tispark.utils.HttpClientUtil
+import com.pingcap.tispark.TiSparkVersion
+import org.apache.spark.sql.SparkSession
+import com.fasterxml.jackson.databind.ObjectMapper
+import com.pingcap.tikv.TiConfiguration
+import com.pingcap.tispark.auth.TiAuthorization
+import org.apache.spark.sql.internal.SQLConf
+import org.slf4j.LoggerFactory
+import scalaj.http.HttpResponse
+import scala.reflect.{ClassTag, classTag}
+import scala.util.matching.Regex
+
+/**
+ * A part of telemetry message about TiSpark version information.
+ */
+object TiSparkTeleInfo {
+
+  private val logger = LoggerFactory.getLogger(getClass.getName)
+  private val pd_address: Option[String] = getPDAddress
+  private val tispark_version: String = getTispark_version
+  private val tidb_version: String = getPd_version
+  private val spark_version: String = org.apache.spark.SPARK_VERSION
+
+  val tiSparkTeleInfo: Map[String, Any] = generateTiSparkTeleInfo()
+
+  /**
+   * Get tiSparkTeleInfo that cantains TiSpark version, TiDB version, Spark Version.
+   *
+   * @return tiSparkTeleInfo
+   */
+  def getTiSparkTeleInfo(): Map[String, Any] = tiSparkTeleInfo
+
+  private def generateTiSparkTeleInfo(): Map[String, Any] = {
+    Map(
+      "tispark_version" -> this.tispark_version,
+      "tidb_version" -> this.tidb_version,
+      "spark_version" -> this.spark_version)
+  }
+
+  def pdAddress: Option[String] = pd_address
+
+  private def getTispark_version: String = {
+    val pattern = new Regex("[0-9]\\.[0-9]\\.[0-9]-SNAPSHOT")
+    val version = TiSparkVersion.version
+    pattern.findFirstIn(version) match {
+      case Some(s) => s
+      case None => "UNKNOWN"
+    }
+  }
+
+  private def getPd_version: String = {
+    val pDStatusOption = requestPD[PDStatus]("/pd/api/v1/status")
+    if (!pDStatusOption.isDefined)
+      "UNKNOWN"
+    else
+      pDStatusOption.get.version
+  }
+
+  private def getCluster_id: String = {
+    val clusterOption = requestPD[Cluster]("/pd/api/v1/cluster")
+    if (!clusterOption.isDefined)
+      "UNKNOWN"
+    else
+      clusterOption.get.id
+  }
+
+  private def requestPD[T: ClassTag](urlPattern: String): Option[T] = {
+    try {
+      if (!pd_address.isDefined) {
+        return Option.empty[T]
+      }
+      val httpClient = new HttpClientUtil
+      var resp: HttpResponse[String] = null
+
+      val conf: TiConfiguration = new TiConfiguration
+      HttpClientUtil.getTLSParam(conf)
+
+      if (conf.isTlsEnable) {
+        val url = "https://" + pd_address.get + urlPattern
+        resp = httpClient.getHttps(url, conf)
+      } else {
+        val url = "http://" + pd_address.get + urlPattern
+        resp = httpClient.get(url)
+      }
+
+      val mapper = new ObjectMapper
+      val entry = mapper.readValue(resp.body, classTag[T].runtimeClass)
+
+      Option(entry.asInstanceOf[T])
+    } catch {
+      case e: Throwable =>
+        logger.info("Failed to get PD version " + e.getMessage)
```

Comment:
it is better to judge if the resp is successful, and get the error message. 
or exception may be thrown in other places without root cause

Reply:
yep, I do a pre-check, judging whether resp is successful or not, and add a warn log here when resp is failede.

Reply:
If the check is not passed, it's better to stop the process.